File: packages/hmssdk_flutter/android/src/main/kotlin/live/hms/hmssdk_flutter/methods/HMSTranscriptionAction.kt
package live.hms.hmssdk_flutter.methods

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import live.hms.hmssdk_flutter.HMSCommonAction
import live.hms.hmssdk_flutter.HMSErrorLogger
import live.hms.hmssdk_flutter.HMSTranscriptExtension
import live.hms.video.sdk.HMSSDK

class HMSTranscriptionAction {

    companion object{

        fun transcriptionAction(call: MethodCall,
                                result: MethodChannel.Result,
                                hmssdk: HMSSDK?){
            when(call.method){
                "start_real_time_transcription" -> {
                    startRealTimeTranscription(call,result,hmssdk)
                }
                "stop_real_time_transcription" -> {
                    stopRealTimeTranscription(call,result,hmssdk)
                }
            }
        }

        /**
         * [startRealTimeTranscription] starts the transcription for everyone in the room
         */
        private fun startRealTimeTranscription(call: MethodCall, result: MethodChannel.Result, hmssdk: HMSSDK?){

            val mode = call.argument<String>("mode") as? String

            mode?.let { _mode ->
                val transcriptionMode = HMSTranscriptExtension.getTranscriptionModeFromString(_mode)
                transcriptionMode?.let {
                    hmssdk?.startRealTimeTranscription(it, HMSCommonAction.getActionListener(result))
                }?:run{
                    HMSErrorLogger.logError("startRealTimeTranscription","Mode is null","NULL Error")
                }
            }
        }

        /**
         * [stopRealTimeTranscription] starts the transcription for everyone in the room
         */
        private fun stopRealTimeTranscription(call: MethodCall, result: MethodChannel.Result, hmssdk: HMSSDK?){
            val mode = call.argument<String>("mode") as? String

            mode?.let { _mode ->
                val transcriptionMode = HMSTranscriptExtension.getTranscriptionModeFromString(_mode)
                transcriptionMode?.let {
                    hmssdk?.stopRealTimeTranscription(it, HMSCommonAction.getActionListener(result))
                }?:run{
                    HMSErrorLogger.returnArgumentsError("mode is a required parameter cannot be null")
                }
            }
        }
    }
}

File: packages/hmssdk_flutter/ios/Classes/Actions/HMSTranscriptionAction.swift
//
//  HMSTranscriptionAction.swift
//  hmssdk_flutter
//
//  Created by Pushpam on 14/06/24.
//

import Foundation
import HMSSDK

class HMSTranscriptionAction{
    
    static func transcriptionActions(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmssdk: HMSSDK?){
        switch call.method{
            case "start_real_time_transcription":
                startRealTimeTranscription(result, hmssdk)
            case "stop_real_time_transcription":
                stopRealTimeTranscription(result, hmssdk)
            default:
                result(FlutterMethodNotImplemented)
        }
    }
    
    private static func startRealTimeTranscription(_ result: @escaping FlutterResult, _ hmssdk: HMSSDK?){

        hmssdk?.startTranscription(){ _, error in
            if let error = error{
                result(HMSErrorExtension.toDictionary(error))
            }else{
                result(nil)
            }
        }
    }
    
    
    private static func stopRealTimeTranscription(_ result: @escaping FlutterResult, _ hmssdk: HMSSDK?){
        hmssdk?.stopTranscription(){_, error in
            if let error = error{
                result(HMSErrorExtension.toDictionary(error))
            }else{
                result(nil)
            }
        }
    }
    
    
}
